lib/common/format.cc

```diff
@@ -0,0 +1,616 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "format.h"
+
+namespace Fortran::common {
+
+template<typename CHAR> CHAR FormatValidator<CHAR>::NextChar() {
+  for (++cursor_; cursor_ < end_; ++cursor_) {
+    if (*cursor_ != ' ') {
+      return toupper(*cursor_);
+    }
+  }
+  cursor_ = end_;
```

Comment:
The loop ends when `cursor_` is equal to `end_`, this is not necessary.

Reply:
Done; thanks.

Reply:
I responded too soon without actually testing this change.   For an unterminated format, cursor_ may be end_+1 after the loop, rather than end_.  This assignment avoids a bogus duplicate error in one of my new tests.  There are other ways to deal with this issue, but this looks like a good way to do that.  I have reinstated the assignments in these two functions.

Reply:
Ah, I see. So `NextChar` could be called with `cursor_` == `end_` and in that case this assignment is required. Could you leave a comment about this?

Reply:
Ok.  I put in a comment, then took it out, since the test catches this case.  I'll put it back.